libflux/src/scanner/scanner.rl

```diff
@@ -8,8 +7,22 @@
 
     include WChar "unicode.rl";
 
-    # For comments on newlines tracking, see below.
-    newline = '\n' @{ push(&nls, fpc - data + 1); no_nls++; };
+    action advance_line {
+        // We do this for every newline we find.
+        // This allows us to return correct line/column for each token
+        // back to the caller.
+        (*cur_line)++;
+        *last_newline = fpc + 1;
+    }
+
+    action advance_line_between_tokens {
+        // We do this for each newline we find in the whitespace between tokens,
+        // so we can record the location of the first byte of a token.
+        last_newline_before_token = *last_newline;
+        cur_line_token_start = *cur_line;
+    }
```

Comment:
OK, now I think I got it, this seems like pretty smart and advanced Ragel use.

So, `advance_line` gets called while you are lexing tokens, so even inside of a single token (e.g. a multiline string), while `advance_line_between_tokens` gets called only outside of token lexing (as if tokens are atomic units).

So `advance_line_between_tokens` saves the last position of the last newline before lexing the next token.  
`advance_line` will update the lines while lexing the next token. When you get the next token you can use `(cur_line_token_start, last_newline_before_token)` for start pos and `(cur_line, last_newline)` for end pos.

I got confused by names, but I guess they are ok 👍 😅

Reply:
Yes, that's exactly right.

Reply:
This is a great way to get around using a linked list. Pretty cool use of Ragel.